Justfile

```diff
@@ -0,0 +1,84 @@
+# FIXME: some of these tests don't work very well on MacOS at the moment. Should fix
+# them at earliest convenience :)
+# Also in addition to this, the `nextest-integration` test is currently disabled on macos
+with_macos_excludes := if os() == "macos" {
+    "--exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
+} else {
+    ""
+}
+nightly_flags := "--features nightly,test_features"
+bash := if os() == "macos" { "/bin/bash" } else { "/usr/bin/bash" }
+
+# all the tests, as close to CI as possible
+test: test-ci test-extra
+
+# only the tests that are exactly the same as the ones in CI
+test-ci: (nextest "stable") (nextest "nightly") check-style
+
+# tests that are as close to CI as possible, but not exactly the same code
+test-extra: check-lychee
+
+# all cargo tests, TYPE is "stable" or "nightly"
+nextest TYPE: (nextest-unit TYPE) (nextest-integration TYPE)
+
+# cargo unit tests, TYPE is "stable" or "nightly"
+nextest-unit TYPE:
+    cargo nextest run \
+        --locked \
+        --workspace \
+        --exclude integration-tests \
+        --cargo-profile dev-release \
+        --profile ci \
+        {{ with_macos_excludes }} \
+        {{ if TYPE == "nightly" { nightly_flags } \
+           else if TYPE == "stable" { "" } \
+           else { error("TYPE is neighter 'nightly' nor 'stable'") } }}
+
+# cargo integration tests, TYPE is "stable" or "nightly"
+[linux]
+nextest-integration TYPE:
+    cargo nextest run \
+        --locked \
+        --package integration-tests \
+        --cargo-profile dev-release \
+        --profile ci \
+        {{ if TYPE == "nightly" { nightly_flags } \
+           else if TYPE == "stable" { "" } \
+           else { error("TYPE is neither 'nightly' nor 'stable'") } }}
+[macos]
+nextest-integration TYPE:
+    @echo "Nextest integration tests are currently disabled on macos!"
+
+# generate a codecov report for RULE
+codecov RULE:
+    #!{{ bash }}
+    set -euxo pipefail
+    source <(cargo llvm-cov show-env --export-prefix)
+    just {{ RULE }}
+    cargo llvm-cov report --profile dev-release --codecov --output-path codecov.json
+
+# style checks
+check-style:
+    python3 scripts/check_nightly.py
+    python3 scripts/check_pytests.py
+    python3 scripts/fix_nightly_feature_flags.py
+    ./scripts/formatting --check
+
+# lychee-based url validity checks
+check-lychee:
+    lychee . {{ if env("GITHUB_TOKEN", "") != "" { "" } else { "-a 429" } }}
```

Comment:
This is not getting run in GHA though, is it? Perhaps consider throwing an `error()` for GHA environment instead, to make it clear to the future readers that this is a local-only step.

Reply:
It’s not running in GHA. But it’s actually useful to have `GITHUB_TOKEN` set locally, if you want to check URL validity locally: github rate limits are such that the second run of `lychee` already hits rate limits. I actually had to have one in order to make the first PR to fix all our dead links :)